tracer/src/Datadog.Trace/Iast/DefaultTaintedMap.cs

```diff
@@ -0,0 +1,244 @@
+    public const int PositiveMask = int.MaxValue;
+    // Periodicity of table purges, as number of put operations. It MUST be a power of two.
+    private const int PurgeCount = 1 << 6;
+    // Bitmask for fast modulo with PURGE_COUNT.
+    private const int PurgeMask = PurgeCount - 1;
+    // Map containing the tainted objects
+    private ConcurrentDictionary<int, ITaintedObject> _map;
+    // Bitmask for fast modulo with table length.
+    private int _lengthMask = 0;
+    // Flag to ensure we do not run multiple purges concurrently.
+    private bool _isPurging = false;
+    private object _purgingLock = new();
+    // Number of hash table entries. If the hash table switches to flat mode, it stops counting elements.
+    private int _entriesCount;
+    /* Number of elements in the hash table before switching to flat mode. */
+    private int _flatModeThreshold;
+
+    public DefaultTaintedMap()
+    {
+        _map = new ConcurrentDictionary<int, ITaintedObject>();
+        _lengthMask = DefaultCapacity - 1;
+        _flatModeThreshold = DefaultFlatModeThresold;
+    }
+
+    /// <summary>
+    /// Gets a value indicating whether flat mode is enabled or not. Once this is set to true, it is not set to false again unless clear() is called.
+    /// </summary>
+    public bool IsFlat { get; private set; } = false;
+
+    /// <summary>
+    /// Returns the ITaintedObject for the given input object.
+    /// </summary>
+    /// <param name="objectToFind">The object that should be found in the map</param>
+    /// <returns>The retrieved tainted object or null</returns>
+    public ITaintedObject? Get(object objectToFind)
+    {
+        if (objectToFind is null)
+        {
+            return null;
+        }
+
+        _map.TryGetValue(IndexObject(objectToFind), out var entry);
+
+        while (entry != null)
+        {
+            if (objectToFind == entry.Value)
+            {
+                return entry;
+            }
+
+            entry = entry.Next;
+        }
+
+        return null;
+    }
+
+    /// <summary>
+    /// Put a new TaintedObject in the dictionary.
+    /// </summary>
+    /// <param name="entry">Tainted object</param>
+    public void Put(ITaintedObject entry)
+    {
+        if (entry is null || (entry.Value as string == string.Empty))
+        {
+            return;
+        }
+
+        var index = Index(entry.PositiveHashCode);
+
+        if (!IsFlat)
+        {
+            // By default, add the new entry to the head of the chain.
+            // We do not control duplicate entries.
+            _map.TryGetValue(index, out var existingValue);
+            entry.Next = existingValue;
+
+            // We only count the entries if we are not in flat mode
+            Interlocked.Increment(ref _entriesCount);
+        }
+
+        // If we flipped to flat mode:
+        // - Always override elements ignoring chaining.
+        // - Stop updating the estimated size.
+
+        _map[index] = entry;
```

Comment:
I don't think this is _technically_ threadsafe? If there's two callers calling `Put` on the same map, and they have the same `index`, then one of the `ITaintedObjects` could be lost`?

e.g.

```csharp
// thread 1
_map.TryGetValue(index, out var existingValue);
entry.Next = existingValue;

// thread 2
_map.TryGetValue(index, out var existingValue);
entry.Next = existingValue;
_map[index] = entry;  // 👈 entry is lost after thread 1 overwrites?

// thread 1
_map[index] = entry;
```

Reply:
Yes, you are right. Adding a new lock for that would solve this. On the other hand, it would cause a penalty in terms of performance. Loosing some taintedObjects from time to time is something that we can afford if this improves the performance. I will talk to the iast team and either solve it or add a comment in the code regarding this.